src/llvm-alloc-opt.cpp

```diff
@@ -0,0 +1,667 @@
+    };
+    // Both `orig_i` and `new_i` should be pointer of the same type
+    // but possibly different address spaces. `new_i` is always in addrspace 0.
+    auto replace_inst = [&] (Instruction *user) {
+        Instruction *orig_i = cur.orig_i;
+        Instruction *new_i = cur.new_i;
+        if (isa<LoadInst>(user) || isa<StoreInst>(user)) {
+            user->replaceUsesOfWith(orig_i, new_i);
+        }
+        else if (auto call = dyn_cast<CallInst>(user)) {
+            if (ptr_from_objref && ptr_from_objref == call->getCalledFunction()) {
+                call->replaceAllUsesWith(new_i);
+                call->eraseFromParent();
+                return;
+            }
+            // remove from operand bundle
+            Type *new_t = new_i->getType();
+            user->replaceUsesOfWith(orig_i, ConstantPointerNull::get(cast<PointerType>(new_t)));
+        }
+        else if (isa<AddrSpaceCastInst>(user) || isa<BitCastInst>(user)) {
+            auto cast_t = PointerType::get(cast<PointerType>(user->getType())->getElementType(),
+                                           0);
+            auto replace_i = new_i;
+            Type *new_t = new_i->getType();
+            if (cast_t != new_t) {
+                replace_i = new BitCastInst(replace_i, cast_t, "", user);
+                replace_i->setDebugLoc(user->getDebugLoc());
+                replace_i->takeName(user);
+            }
+            push_frame(user, replace_i);
+        }
+        else if (auto gep = dyn_cast<GetElementPtrInst>(user)) {
+            SmallVector<Value *, 4> IdxOperands(gep->idx_begin(), gep->idx_end());
+            auto new_gep = GetElementPtrInst::Create(gep->getSourceElementType(),
+                                                     new_i, IdxOperands,
+                                                     gep->getName(), gep);
+            new_gep->setIsInBounds(gep->isInBounds());
+            new_gep->takeName(gep);
+            copyMetadata(new_gep, gep);
+            push_frame(gep, new_gep);
+        }
+        else {
+            abort();
+        }
+    };
+
+    while (true) {
+        replace_inst(cast<Instruction>(*cur.orig_i->user_begin()));
+        while (cur.orig_i->use_empty()) {
+            finish_cur();
+            if (stack.empty())
+                return;
+            cur = stack.back();
+            stack.pop_back();
+        }
+    }
+}
+
+bool AllocOpt::isSafepoint(Instruction *inst)
+{
+    auto call = dyn_cast<CallInst>(inst);
+    if (!call)
+        return false;
+    if (isa<IntrinsicInst>(call))
+        return false;
+    if (auto callee = call->getCalledFunction()) {
+        // Known functions emitted in codegen that are not safepoints
+        if (callee == ptr_from_objref || callee->getName() == "memcmp") {
+            return false;
+        }
+    }
+    return true;
+}
+
+bool AllocOpt::runOnFunction(Function &F)
+{
+    if (!alloc_obj)
+        return false;
+    std::map<CallInst*,size_t> allocs;
```

Comment:
std::map is generally discouraged in LLVM passes because of nondeterministic iteration order. Can you use a data structure with deterministic iteration?